docs(material/list): switch examples and demo to MDC

crisbeto opened the pull request:
Switches the list live examples and demo to use MDC. Also updates the docs to reflect the latest API.

devversion commented on src/material/list/list.md:
```diff
@@ -24,18 +24,18 @@ Simple navigation lists can use the `mat-list-item` attribute on anchor tag elem
   <mat-list-item *ngFor="let link of links">
-     <a matLine href="...">{{ link }}</a>
-     <button mat-icon-button (click)="showInfo(link)">
+     <a matListItemLine href="...">{{ link }}</a>
```
I wonder if we need to clarify the API a bit more, of should that be a follow-up?

Every list item must have a title — it cannot use lines without an explicit title. Maybe also http://go/angular-mdc/migration#list is helpful.

```suggestion
     <a matListItemTitle href="...">{{ link }}</a>
```

crisbeto replied:
Changed the example to have a title. Do we have validation at runtime that it has a title? If it does, I don't think that we need to call it out specifically.

devversion replied:
Yeah, we have some sanity checks in: https://github.com/angular/components/blob/main/src/material/list/list-base.ts#L323. I think it would be still nice at some point to talk about the new API more in detail — like the support for text wrapping and the explicit `lines` attribute.

devversion reviewed (approved):
LGTM for the sources

devversion reviewed (approved):
LGTM if you want to proceed with this for now.

mmalerba commented on src/material/list/list.md:
```diff
@@ -24,18 +24,18 @@ Simple navigation lists can use the `mat-list-item` attribute on anchor tag elem
-     <a matLine href="...">{{ link }}</a>
-     <button mat-icon-button (click)="showInfo(link)">
+     <a matListItemTitle href="...">{{ link }}</a>
+     <button mat-icon-button (click)="showInfo(link)" matListItemMeta>
```
can we add some docs about `matListItemTitle` and `matListItemMeta`? I don't think they had an equivalent in the old version, so they're not documented

mmalerba commented on src/material/list/list.md:
````diff
@@ -24,18 +24,18 @@ Simple navigation lists can use the `mat-list-item` attribute on anchor tag elem
 ```html
 <mat-nav-list>
-   <a mat-list-item href="..." *ngFor="let link of links"> {{ link }} </a>
+  <a mat-list-item href="..." *ngFor="let link of links">{{ link }}</a>
````
we should mention the new `activated` input for nav-lists somewhere

crisbeto commented:
@devversion @mmalerba I've addressed the feedback. I also removed the section about dense lists since we don't have an equivalent attribute now and I'm guessing that it'll be handled using the density Sass API.

devversion commented on src/material/list/list.md:
````diff
@@ -1,21 +1,59 @@
+```
+
+To activate text wrapping, the `lines` input has to be set on the `<mat-list-item>` indicating the
+number of lines of text.
````
Do we want to add some examples for this? There is one clarification needed I think:

1. Unscoped content (stuff not wrapped by e.g. `matListItemTitle/Line`) will take up the remaining line space.
2. Text inside `matListItemTitle` or `matListItemLine` *never* can wrap/never takes up the remaining line space.
3. If lines is not set, the lines are automatically inferred based on how many lines, and an extra line if there is unscoped content.

All of this can be a follow-up. I'm happy to work on this too — but I think we should explain this more.

crisbeto replied:
I'm confused about the whole scoped/unscoped lines and text wrapping. I'm trying the following:

```
<mat-list>
  <mat-list-item *ngFor="let message of messages; last as last">
    <img matListItemAvatar [src]="message.image">
    <div matListItemTitle>{{message.from}}</div>
    <div>
      <span>{{message.subject}} -- </span>
      <span>{{message.message}}</span>
    </div>
    <mat-divider inset *ngIf="!last"></mat-divider>
  </mat-list-item>
</mat-list>
```

Which produces the following with broken spacing:
<img width="822" alt="image" src="https://user-images.githubusercontent.com/4450522/186670096-0955789c-256e-40dc-904a-f48141a04b9a.png">

It looks better if I remove the `div` around the two `span` elements, but the lines still don't wrap, even if I add a `lines="2"` to the list item.

devversion replied:
**Unscoped content:** Text content outside of any explicit `matListItemTitle` or `matListItemLine`. Open for better names here— this is the term I used in the design document.

Yeah, the use of `div` for the title, or for the second line is unexpected as it introduces additional spacing. If you switch all divs to `span`, then it works as expected.

If you do not specify any lines, then there is a title + line (the content outside of any explicit line is picked up automatically as a line). If you would set `lines=3`, then the unscoped content would actually be acquire the remaining space. The content of the second line can then wrap and take up the remaining space. Without `lines`—it would just be a single line + title and neither would wrap (and just show an ellipsis). There quite a few examples in the dev-app `mdc-list` demo.

e.g.

```html
<mat-list-item>
  <h3 matListItemTitle>Title</h3>
  Unscoped content — Will take up the remaining space— since there is no explicit 
  `line` count. It will just be treated as an additional line— like with `span matListItemLine`
</mat-list-item>
```

```html
<mat-list-item lines="3">
  <h3 matListItemTitle>Title</h3>
  Unscoped content — Will take up the remaining space  — in this case the list has explicit 3 lines, so
  without the title, the unscoped content will span over two lines (and wrap).
</mat-list-item>
```

Note: Even if the content in the `lines=3` case is too short, the list item acquires the space of three lines. Consumers need to chose beforehand if they want to reserve the space for three lines or not. This came from MDC. A list never automatically grows if the content would exceed.

devversion replied:
Let me know if that helps, or if we should go through some examples over VC/chat. Happy to help

crisbeto replied:
As discussed over DM, I'll merge the changes as they are now so the migration isn't blocked. The docs content can be iterated over afterwards.
